src/glview/VertexStateManager.cc

```diff
@@ -0,0 +1,61 @@
+#include "VertexStateManager.h"
+
+void VertexStateManager::initializeSize(size_t vertices_size) {
+  std::vector<GLuint> placeholder = std::vector<GLuint>();
+  size_t placeholder_index = 0;
+  initializeSizeHelper(vertices_size, false, placeholder, placeholder_index);
+}
+
+void VertexStateManager::initializeSize(size_t vertices_size, std::vector<GLuint> & vbos, size_t & vbo_index) {
+  initializeSizeHelper(vertices_size, true, vbos, vbo_index);
+}
+
+void VertexStateManager::initializeSizeHelper(size_t vertices_size, bool multiple_vbo, std::vector<GLuint> & vbos, size_t & vbo_index) {
+  if (Feature::ExperimentalVxORenderersDirect.is_enabled() || Feature::ExperimentalVxORenderersPrealloc.is_enabled()) {
+    size_t elements_size = 0;
+    if (Feature::ExperimentalVxORenderersIndexing.is_enabled()) {
+      if (multiple_vbo) {
+        vertex_array->elementsVBO() = vbos[vbo_index++];
+      }
+      if (vertices_size <= 0xff) {
+          vertex_array->addElementsData(std::make_shared<AttributeData<GLubyte, 1, GL_UNSIGNED_BYTE>>());
+      } else if (vertices_size <= 0xffff) {
+          vertex_array->addElementsData(std::make_shared<AttributeData<GLushort, 1, GL_UNSIGNED_SHORT>>());
+      } else {
+          vertex_array->addElementsData(std::make_shared<AttributeData<GLuint, 1, GL_UNSIGNED_INT>>());
+      }
+      elements_size = vertices_size * vertex_array->elements().stride();
+      vertex_array->elementsSize(elements_size);
+    }
+    vertices_size *= vertex_array->stride();
+    vertex_array->verticesSize(vertices_size);
+
+    GL_TRACE("glBindBuffer(GL_ARRAY_BUFFER, %d)", vertex_array->verticesVBO());
+    GL_CHECKD(glBindBuffer(GL_ARRAY_BUFFER, vertex_array->verticesVBO()));
+    GL_TRACE("glBufferData(GL_ARRAY_BUFFER, %d, %p, GL_STATIC_DRAW)", vertices_size % (void *)nullptr);
+    GL_CHECKD(glBufferData(GL_ARRAY_BUFFER, vertices_size, nullptr, GL_STATIC_DRAW));
+    if (Feature::ExperimentalVxORenderersIndexing.is_enabled()) {
+      GL_TRACE("glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, %d)", vertex_array->elementsVBO());
+      GL_CHECKD(glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, vertex_array->elementsVBO()));
+      GL_TRACE("glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, %d, %p, GL_STATIC_DRAW)", elements_size % (void *)nullptr);
+      GL_CHECKD(glBufferData(GL_ELEMENT_ARRAY_BUFFER, elements_size, nullptr, GL_STATIC_DRAW));
+    }
+  } else if (Feature::ExperimentalVxORenderersIndexing.is_enabled()) {
+    if (multiple_vbo) {
+      vertex_array->elementsVBO() = vbos[vbo_index++];
+    }
+    vertex_array->addElementsData(std::make_shared<AttributeData<GLuint, 1, GL_UNSIGNED_INT>>());
+  }
+}
+void VertexStateManager::addColor(Color4f last_color) {
```

Comment:
This `addColor` version won't work as a replacement inside `CGALRenderer` once it's time to merge the work done in #4330. That's because it introduces a new attribute for the shader `color_marked`. However, that may change (see the [comment here](https://github.com/openscad/openscad/pull/4330/commits#discussion_r1154840778))

Reply:
I am interested in the color_marked attribute. When can an user see an item with color_marked ?

Reply:
@gsohler The `color_marked` attribute was actually a workaround so that `PolySet` can be used to display the green-colored faces that happen when a `difference()` operation gets used. This `color_marked` attribute does not currently let the user programmatically select a custom set of faces to be rendered with a different color. But maybe you could adapt some of my code from #4330 to allow that type of highlighting behavior if it's what you want.

Note that this PR #4675 is refactoring only and doesn't introduce anything related to `color_marked`. I put that note here because the now-closed #4330 from last year's GSoC interferes with this change, so it is something I'll need to figure out myself later.